sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/IndentingXMLStreamWriter.scala

```diff
@@ -0,0 +1,208 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.catalyst.xml
+
+import javax.xml.namespace.NamespaceContext
+import javax.xml.stream.XMLStreamWriter
+
+import scala.collection.mutable.Stack
+
+class IndentingXMLStreamWriter(wr: XMLStreamWriter) extends XMLStreamWriter {
+  private sealed trait WriterState
+  private case object NO_NEWLINE_ON_END extends WriterState
+  private case object DO_NEWLINE_ON_END extends WriterState
+
+  private val writer = wr
+
+  private var indentStep = "  "
+  private var depth = 0
+
+  private var state: WriterState = NO_NEWLINE_ON_END
+  private val stateStack = Stack[WriterState]()
```

Comment:
Also this is pretty well-known algorithm to use stack + indentation, which doesn't require a license issue up to my best knowledge. I think it's fine to reopen this.